src/ngraph/pass/algebraic_simplification.cpp

```diff
@@ -419,6 +432,32 @@ static bool simplify_log(shared_ptr<Node> n)
     return false;
 }
 
+// optimizes `gather->shapeof` into `shapeof->gather` for 0D gather indices
+static bool simplify_gather_shapeof(shared_ptr<Node> node)
+{
+    auto shapeof = as_type_ptr<op::v0::ShapeOf>(node);
+    auto gather = as_type_ptr<op::v1::Gather>(shapeof->input_value(0).get_node_shared_ptr());
+    if (!gather)
+    {
+        return false;
+    }
+    auto indices = as_type_ptr<op::Constant>(gather->input_value(1).get_node_shared_ptr());
+    if (!indices || indices->get_shape() != Shape{} ||
```

Comment:
1. We could support `Shape{1}` case too.
2. Why do we require indices to be constant and have static shape? Static rank check is enough, isn't it?
3. Why we use numeric_limits instead of AXIS_NOT_SET_VALUE?

Reply:
only 0D indices ( Shape{} ) result in axis dimension being removed.  Will try make other changes.